DOCS-9094: Fixed when only a single flow instances is allowed (#1645)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/scheduler-concept.adoc

include::_attributes.adoc[]
endif::[]

> The Scheduler is a flow event source that is configured to trigger at certain times to execute the Scheduler's flow. A Scheduler is either configured with a fixed repeating triggering *frequency* (such as every second), or with a more general cron *expression*. A cron expression can specify one or more specific dates and times, but more often specifies recurring dates, times, or intervals (for example the cron expression `* * * ? * *` means to trigger every second). 

Reviewer: ```suggestion
The Scheduler component is an event source in a flow that initiates execution of that flow at configured times. You can configure a fixed and repeating frequency (such as every second) or a cron expression. A cron expression can specify one or more specific dates and times, but more commonly, the expression specifies recurring dates, times, or intervals. For example, the cron expression `* * * ? * *` configures a Scheduler to start its flow every second. 
```

> The Scheduler is a flow event source that is configured to trigger at certain times to execute the Scheduler's flow. A Scheduler is either configured with a fixed repeating triggering *frequency* (such as every second), or with a more general cron *expression*. A cron expression can specify one or more specific dates and times, but more often specifies recurring dates, times, or intervals (for example the cron expression `* * * ? * *` means to trigger every second). 

Reviewer: ```suggestion
The Scheduler component is an event source in a flow that initiates execution of that flow at configured times. You can configure a fixed and repeating frequency (such as every second) or a cron expression. A cron expression can specify one or more specific dates and times, but more commonly, the expression specifies recurring dates, times, or intervals. For example, the cron expression `* * * ? * *` configures a Scheduler to start its flow every second. 
```

Reviewer: moved change to new PR: https://github.com/mulesoft/docs-mule-runtime/pull/1650

> The flow can be configured to only allow the Scheduler event source to execute one flow instance at a time, or can allow concurrent flow instances to execute. 

Reviewer: ```suggestion
You can configure the Flow component to run concurrent instances of the flow or to prevent the Scheduler from starting more than one instance of the flow at a time. 
```

> The flow can be configured to only allow the Scheduler event source to execute one flow instance at a time, or can allow concurrent flow instances to execute. 

Reviewer: ```suggestion
A `maxConcurrency` setting on the Flow component affects the flow's Scheduler. The setting limits the number of instances of the flow that can run at the same time. Setting the value to `1` prevents the Scheduler from triggering more than one execution of the flow at a time. 
```

Reviewer: moved change to new PR: https://github.com/mulesoft/docs-mule-runtime/pull/1650

Consider the following when you implement a Scheduler in your Mule application:

* Schedulers use the same timezone as the machine on which Mule is running. However, if an application is running in CloudHub, the Scheduler conforms to the UTC timezone, regardless of the geographic region in which the application is running.
// TODO: MULE-14438
* If xref:execution-engine.adoc#backpressure[back-pressure] is triggered because no resources are available at the time of the trigger, that trigger is skipped.
//Enhancement request for this: MULE-14930

> * If the flow is configured with `maxconcurrency` set to 1, then while a triggered flow is still executing, subsequent Scheduler triggers cause the Scheduler's back-pressure to silently skip the trigger and no new flow instances are executed. 

Reviewer: ```suggestion
* If the `maxconcurrency` setting of a Flow component is set to `1`, no new flow instances will execute while the flow is executing. The Scheduler's back-pressure silently skips triggers while the flow is executing. 
```

Reviewer: This suggested change was replaced with another below.

> * If xref:execution-engine.adoc#backpressure[back-pressure] is triggered because no resources are available at the time of the trigger, that trigger is skipped. In this case a *MULE:FLOW_BACK_PRESSURE* error is thrown indicating the flow cannot accept new events at this time. In Anypoint Studio, this error is only logged once the first time it occurs. You can capture log messages each time the Scheduler is triggered, including events that do not execute the flow, by enabling debug level logging in the Mule application.  

Reviewer: ```suggestion
* If xref:execution-engine.adoc#backpressure[back pressure] occurs because Mule lacks sufficient resources to process an event in the flow, the application skips the scheduled trigger for the event and throws a MULE:FLOW_BACK_PRESSURE error. This error indicates that the flow cannot accept new events at that time. Anypoint Studio logs this error only once, the first time it occurs. 
+
To capture log messages each time the Scheduler attempts to trigger an event, you can enable the xref:logging-in-mule#change-the-log-level[DEBUG] logging level in the Mule application. 
```

Reviewer: moved change to new PR: https://github.com/mulesoft/docs-mule-runtime/pull/1650

> * If the `maxConcurrency` attribute of a Flow component is set to `1`, then only one flow instance can execute at a time in the Mule runtime. In this case, The Mule runtime prevents new flow instances from executing while the first flow instance is still executing. Instead, the Scheduler's back-pressure does *not* queue up additional events. Instead, the flow skips new events created by the triggered Scheduler while the first flow is still executing. Once the current flow terminates, the next event triggered by the Scheduler will execute a new flow instance, and then the same blocking and event skipping pattern repeats. 

Reviewer: ```suggestion
* If you set the `maxConcurrency` attribute of a Flow component to `1`, only one instance of the flow can execute at a time. When an instance is running, the application skips and avoids queuing any new events that are triggered by the Scheduler. Once that instance of the flow terminates and the Scheduler triggers an event, a new instance of the flow executes, and the same blocking and event-skipping pattern occurs. 
+
See also, `disallowConcurrentExecution` in xref:scheduler-xml-reference.adoc#scheduler-properties[<scheduler> Properties]. 
```

Reviewer: moved change to new PR: https://github.com/mulesoft/docs-mule-runtime/pull/1650

> * In a multi-worker CloudHub deployment, or in a custoemr-hosted Mule runtime cluster, the Scheduler runs (and then executes the flow) only on the primary node (that is, only in one particular Mule runtime).

Reviewer: ```suggestion
* In a multi-worker CloudHub deployment or in a customer-hosted cluster of Mule runtime engines, the Scheduler runs and executes the flow on the primary node only. That is, it runs on one specific Mule runtime engine and no others.
```

Reviewer: moved change to new PR: https://github.com/mulesoft/docs-mule-runtime/pull/1650

* Schedulers use the same timezone as the machine on which Mule is running. However, if an application is running in CloudHub, the Scheduler conforms to the UTC timezone, regardless of the geographic region in which the application is running.
// TODO: MULE-14438

> * If xref:execution-engine.adoc#backpressure[back-pressure] is triggered because no resources are available at the time of the trigger, that trigger is skipped. In this case a *MULE:FLOW_BACK_PRESSURE* error is thrown indicating the flow cannot accept new events at this time.  

Reviewer: ```suggestion
* If xref:execution-engine.adoc#backpressure[back-pressure] is triggered because no resources are available at the time of the trigger, the application skips the trigger and throws a MULE:FLOW_BACK_PRESSURE error. The error indicates that the flow cannot accept new events at that time.  
```

Reviewer: moved change to new PR: https://github.com/mulesoft/docs-mule-runtime/pull/1650

//Enhancement request for this: MULE-14930

> * If the flow is configured with `maxconcurrency` set to 1, then while a triggered flow is still executing, subsequent Scheduler triggers cause the Scheduler's back-pressure to skip the trigger and no new flow instances are executed. This guarantees that only one instance of the flow is executing in the Mule runtime at a particular time. 

Reviewer: ```suggestion
* If the `maxconcurrency` attribute of a Flow component is set to `1`, then only one flow instance can execute at a time in the Mule runtime. In this case, The Mule runtime prevents new flow instances from executing while the first flow instance is still executing. Instead, the Scheduler's back-pressure does *not* queue up additional events. Instead, the flow skips new events created by the triggered Scheduler while the first flow is still executing. Once the current flow terminates, the next event triggered by the Scheduler will execute a new flow instance, and then the same blocking and event skipping pattern repeats. 
```

== Fixed Frequency

You can use the following parameters to set a Scheduler to trigger and execute its flow at a regular intervals.

[%header,cols="1,2,1"]
|===
|Attribute |Description |Default
|*Frequency*
|The frequency at which the Scheduler triggers execution of its flow. You select the unit of time in the *Time unit* field. Setting the value to `0` or a negative value causes the default value to be used.
|
1000

| *Start Delay*

> | The amount of time to wait before triggering execution of its flow for the first time after the application is started. This value is expressed in the same unit of time as the frequency.

Reviewer: ```suggestion
| The amount of time to wait before triggering the execution of its flow for the first time after the application starts. This value is expressed in the same unit of time as the frequency.
```

== Cron Expressions

> Cron is a widely used standard for describing time and date information and repeating intervals.

Reviewer: ```suggestion
Cron is a widely used standard for describing date-time information and repeating intervals.
```

The Scheduler keeps track of every second and creates a Mule event when the

> Quartz Cron expression matches your time-date setting. You can configure the Scheduler to trigger

Reviewer: ```suggestion
Quartz cron expression matches your date-time setting. You can configure the Scheduler to trigger
```

just once or at regular intervals.

> A *date-time expression* consists of six required settings and can include the

Reviewer: ```suggestion
A date-time expression consists of six required settings and can include the
```

> A *date-time expression* consists of six required settings and can include the

Reviewer: ```suggestion
A date-time expression consists of six required settings and can include the
```

Reviewer: note that things in bold do not get translated

Reviewer: moved change to new PR: https://github.com/mulesoft/docs-mule-runtime/pull/1650

//source info: +http://www.quartz-scheduler.org/documentation/quartz-2.x/tutorials/crontrigger.html+

> This example logs the message "hello" every second, and only allows one flow instance to execute at a time:

Reviewer: ```suggestion
This example logs the message "hello" every second and allows only one instance of the flow to execute at a time:
```

Reviewer: moved change to new PR: https://github.com/mulesoft/docs-mule-runtime/pull/1650

//source info: +https://docs.oracle.com/javase/8/docs/api/java/time/ZoneId.html+

> == Debugging a Scheduler

Reviewer: ```suggestion
== Stop or Start a Scheduler in Debug Mode
```

Reviewer: moved change to new PR: https://github.com/mulesoft/docs-mule-runtime/pull/1650

> When you debug a Mule application in Anypoint Studio, each Scheduler event source appears with an arrow icon underneath to remind you that the Scheduler can be stopped and started while the Mule application is being debugged. To stop or start the Scheduler event source, right click on the Scheduler and select Stop Scheduler or Start Scheduler from the Scheduler's right-click menu.  

Reviewer: ```suggestion
When you run a Mule application in Anypoint Studio using debug mode, an arrow icon appears at the bottom of the Scheduler component to indicate that you can start or stop it from the Scheduler's right-click menu.

By default, the Scheduler runs when in debug mode. If you place a break point on a message processor in the flow, the Scheduler component will start the flow but will not run again until you remove or skip all break points from the flow.
```

Reviewer: moved change to new PR: https://github.com/mulesoft/docs-mule-runtime/pull/1650

== Debugging

> In the Studio *Debug Mode* view, the Mule app disables triggering of Schedulers. When debugging, you can

Reviewer: ```suggestion
In the Studio *Debug Mode* view, a Mule application disables triggers on Scheduler components. When debugging, you can
```

Reviewer: We need to verify whether it's true that Studio disables triggers when in debug mode.

Author: It does not disable the Scheduler

Reviewer: Yep, I'm finding that it only stops if you stop the flow by adding a break point or if you actively select Stop Scheduler from the right-click menu on the Scheduler...

> instead trigger a Scheduler component when debugging by clicking the green arrow icon

Reviewer: ```suggestion
instead trigger a Scheduler component by clicking the green arrow icon
```
